### PR TITLE
prevent mixup of account timelines

### LINKED ISSUE #2663: Old toots stuck on top of home timeline
Hi,

After going on my profile and scrolling to an old toot (knew I tooted it 2022-07-25, so just scrolled there), my home timeline now has toots from roughly that period stuck on top - fresh toots starting ~8 screen sizes below them.

The stuck ones seem to be non-contiguous, I think I should have had more toots in home as it shows. The fresh ones are refreshed, but stay at their place below the stuck ones.

I tried purging the app cache and switching tabs and accounts, without success.

This is pretty obscure, not sure what other info could be useful here or how to reproduce it.. some slightly rusted android dev skills are available, so feel free to ask for any adb thing (or something else) that might help diagnose the problem :)

* * * *
- Tusky Version: 19.0
- Android Version: 12 (SQ3A.220705.003)
- Android Device: Pixel 6 Pro
- Mastodon instance (if applicable): mastodon.social

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate (browsed issues for some pages).

### LINKED ISSUE #4148: With multiple accounts, sometimes the wrong home timeline is shown
I have experienced a problem with multiple accounts set up in Tusky where after switching from account 2 (pixelfed) to account 1 (mastodon), the home timeline of account 2 (pixelfed) would be shown and even after switching back and forth between accounts it would stay that way. Only way to fix it seems to be deleting the account showing the wrong timeline from Tusky and adding it again.

I have seen this several times recently (since 24 beta and now in 24), not sure if I had it before. I have not found a way to reproduce it yet. I don't know if the networks are relevant, I added them just in case.

* * * *
- Tusky Version: 24
- Android Version: Lineageos 20 (Android 13)

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.

### LINKED ISSUE #4402: Multiple accounts bug: boosts from one account's home timeline appear in another's
### Detailed description

I have multiple accounts connected to Tusky.
Account 1 follows maloki. Account 2 does not.
multiple boosts from maloki appeared today in Account 2's home timeline.
Screenshot_20240428-213752_Tusky
trying to view the post or the boosting account results in an error:
Screenshot_20240428-212743_Tusky
Screenshot_20240428-213803_Tusky

### Steps to reproduce the problem

1. Connect Tusky to multiple accounts
2. Have Account 1 follow Some User.
3. Browse the home timeline of Account 2
4. boosts from Some Account appear in Account 2's timeline

### Debug information

Tusky 25.0 beta 1

Device:

samsung SM-A125F
Android version: 12
SDK version: 31

Account:

@yuvalne@tooot.im
Version: 4.2.8

### LINKED ISSUE #4554: Old posts from other account shown on timeline, before posts from correct account.
### Detailed description

I have several accounts (let's call them A, B, C ... F) in Tusky on Android. Several of these are on Mastodon servers, and others are on GoToSocial servers. One of those GoToSocial servers, constellation.home.nora.codes, has extremely limited federation; the posts on that instance should exist on 2 other servers, exclusively (mutual federation, but none of us federate anywhere else). Let's call that account F.

However, sometimes when I add accounts to Tusky, some other accounts - say, B and D - show 30 - 32 old (May of this year) posts from that instance in their home timeline. Refreshing and clearing cache do nothing; logging out of F does nothing; logging out of B and D and back in *sometimes* fixes the issue. Reinstalling the app sometimes fixes the issue. Scrolling down shows posts from B or D's timeline below the 30 errant posts.

There is no way the instances for accounts B and D have any knowledge of posts from constellation.home.nora.codes, so I think this is purely a client issue.

### Steps to reproduce the problem

_No response_

### Debug information

Google Pixel 6a
Android version: 13
SDK version: 33
Tusky 25.2 (but has been happening since months ago)
constellation.home.nora.codes (GoToSocial) and types.pl (Mastodon)

### LINKED ISSUE #4567: Tusky 26 showing notifications tab from wrong account
### Detailed description

I'm using Tusky with three different Mastodon accounts (all different instances). Since the update to v26 (F-droid), the notifications tab for account_A@imaginair.es shows notifications for the latest other account to be opened.

I have tried to solve the issue by force stopping Tusky and clearing the app cache, same result. Checked imaginair.es which can sometimes be bottlenecked, it's up so shouldn't be the issue.

I imagine this has to do with the new notifications cache?

### Steps to reproduce the problem

1. Use Tusky v26 with multiple accounts
2. Switch between accounts
3. Open notifications tab

Expected: see notifications for the current account.
Issue: notifications for the latest used account shown instead.

### Debug information

Tusky 26.0

Device:

OnePlus ONEPLUS A6003
Android version: 10
SDK version: 29

Account:

@REDACTED@imaginair.es
Version: 4.3.0-alpha.3+glitch

### PR DESCRIPTION
**Pull request by connyduck**
This does 2 things:

- Removes `AccountSwitchInterceptor`, the main culprit for the bug. APIs can no longer change their base url after they have been created. As a result they are not Singletons anymore.
- Additionally, I refactored how MainActivity handles Intents to make it less likely to have multiple instances of it active.

Here is how I could reliably reproduce the bug:

- Be logged in with account A and B
- Write a post with account A, cancel it before it sends (go into flight mode for that)
- Switch to account B
- Open the "this post failed to send" notification from account A, drafts will open
- Go back. You are in the MainActivity of account A, everything seems fine.
- Go back again. You are in the old, now broken MainActivity of account B. It uses the database of account B but the network of account A. Refreshing will show posts from A.

closes #4567 
closes #4554
closes #4402 
closes #4148
closes #2663
and possibly #4588

### CODE REVIEW

File: app/src/main/java/com/keylesspalace/tusky/di/NetworkModule.kt [hunk 1]
```diff
@@ -66,6 +68,18 @@ object NetworkModule {
...
+    fun providesDefaultPort(): Int {
+        return 443
+    }
+
+    @Provides
+    @Named("defaultScheme")
+    fun providesDefaultScheme(): String {
+        return "https://"
+    }
+
```
Comment (connyduck) [inline app/src/main/java/com/keylesspalace/tusky/di/NetworkModule.kt:82]: This is kinda weird, but the only way I found to have tests run on a random port without https.
Better ideas very welcome!

File: app/src/main/java/com/keylesspalace/tusky/di/NetworkModule.kt [hunk 2]
```diff
@@ -125,43 +138,100 @@ object NetworkModule {
...
+                }
+
+                if (currentAccount != null && request.url.host == currentAccount.domain) {
+                    request = request.newBuilder()
+                        .header("Authorization", "Bearer %s".format(currentAccount.accessToken))
+                        .build()
+                }
+
+                print(originalRequest)
+                print(request)
```
Comment (Tak) [inline app/src/main/java/com/keylesspalace/tusky/di/NetworkModule.kt:227]: Intended?

**Review by Tak — APPROVED** [review]

File: app/src/main/java/com/keylesspalace/tusky/di/NetworkModule.kt (hunk 2 above)
Comment (connyduck) [reply app/src/main/java/com/keylesspalace/tusky/di/NetworkModule.kt]: nope, thx

File: app/src/main/java/com/keylesspalace/tusky/MainActivity.kt [hunk 3]
```diff
@@ -363,6 +363,11 @@ class MainActivity : BottomSheetActivity(), ActionButtonActivity, MenuProvider {
         }
     }
 
+    override fun onDestroy() {
+        super.onDestroy()
+        cacheUpdater.stop()
```
Comment (charlag) [inline app/src/main/java/com/keylesspalace/tusky/MainActivity.kt:368]: I think it should be stopped first and then the rest but idk

File: app/src/main/java/com/keylesspalace/tusky/di/NetworkModule.kt (hunk 1 above)
Comment (charlag) [reply app/src/main/java/com/keylesspalace/tusky/di/NetworkModule.kt:82]: could we provide the whole URL depending on the current account?

File: app/src/main/java/com/keylesspalace/tusky/di/NetworkModule.kt [hunk 4]
```diff
@@ -125,43 +138,97 @@ object NetworkModule {
...
+        val currentAccount = accountManager.activeAccount
         return retrofit.newBuilder()
-            .client(longTimeOutOkHttpClient)
+            .apply {
+                if (currentAccount != null) {
+                    baseUrl("$defaultScheme${currentAccount.domain}:$defaultPort")
+                }
+            }
+            .callFactory { originalRequest ->
+                var request = originalRequest
```
Comment (charlag) [inline app/src/main/java/com/keylesspalace/tusky/di/NetworkModule.kt:208]: i think that's a lot of code for DI module tbh, it shows in test that you ask it from DI which seems bit off to me, but not critical ofc, we can refactor this later

**Comment by connyduck** [comment]
Ok, I perf tested this and found no problems. There are a few KBs more memory used, activity start time is the same as before.

I fixed a bug where sometimes the wrong api instance was used because the containing object was a Singleton and the api was not recreated after account switching.

And I refactored a bit to have less code in the NetworkModule and make the testing simpler.

Please have another look.

**Review by Tak — APPROVED** [review]

**Pull request merged**